### PR TITLE
bpo-43923: Add support of generic typing.NamedTuple

### LINKED ISSUE #88089: Can't create generic NamedTuple as of py3.9
BPO | [43923](https://bugs.python.org/issue43923)
--- | :---
Nosy | @gvanrossum, @rhettinger, @ericvsmith, @serhiy-storchaka, @graingert, @ilevkivskyi, @dlukes, @JelleZijlstra, @FHTMitchell, @sobolevn, @Fidget-Spinner, @AlexWaygood, @juliusgeo
PRs | <li>python/cpython#31679</li><li>python/cpython#31779</li><li>python/cpython#31781</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2021-04-23.16:22:56.566>
labels = ['type-bug', '3.9', '3.10', '3.11']
title = "Can't create generic NamedTuple as of py3.9"
updated_at = <Date 2022-03-10.21:38:13.215>
user = 'https://github.com/FHTMitchell'
```

bugs.python.org fields:
```python
activity = <Date 2022-03-10.21:38:13.215>
actor = 'behackett'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = []
creation = <Date 2021-04-23.16:22:56.566>
creator = 'FHTMitchell'
dependencies = []
files = []
hgrepos = []
issue_num = 43923
keywords = ['patch']
message_count = 25.0
messages = ['391705', '406083', '406124', '414541', '414557', '414558', '414559', '414563', '414567', '414568', '414570', '414584', '414585', '414589', '414592', '414593', '414705', '414709', '414754', '414791', '414792', '414815', '414816', '414817', '414830']
nosy_count = 16.0
nosy_names = ['gvanrossum', 'rhettinger', 'eric.smith', 'python-dev', 'serhiy.storchaka', 'graingert', 'levkivskyi', 'dlukes', 'JelleZijlstra', 'behackett', 'FHTMitchell', 'Steven Silvester', 'sobolevn', 'kj', 'AlexWaygood', 'juliusgeo']
pr_nums = ['31679', '31779', '31781']
priority = 'normal'
resolution = None
stage = 'patch review'
status = 'open'
superseder = None
type = 'behavior'
url = 'https://bugs.python.org/issue43923'
versions = ['Python 3.9', 'Python 3.10', 'Python 3.11']
```

</p></details>

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
Closes #88089.

### CODE REVIEW

**Comment by JelleZijlstra** [comment]
Thanks for implementing this, but I think it needs some coordination with static type checkers to see if they have concerns. I know mypy's architecture makes generic namedtuples a little hard to support.

**Comment by AlexWaygood** [comment]
> Thanks for implementing this, but I think it needs some coordination with static type checkers to see if they have concerns. I know mypy's architecture makes generic namedtuples a little hard to support.

I still disagree on this — there's no obligation on type checkers to support generic NamedTuples if this is merged. This does, however, open up the possibility of them being able to support that feature if they ever decide they'd like to in the future.

**Comment by JelleZijlstra** [comment]
I probably agree, but I want to give type checker maintainers a chance to say "don't do this, it will introduce a giant soundness hole". I'll ask about it in the summit in a few hours.

File: Lib/typing.py
Comment (gvanrossum) [inline Lib/typing.py]: I would honestly just violate the line length limit and keep the entire error message string on one line.

**Review by gvanrossum — COMMENTED** [review]
Other than one nit I would be fine if this goes into 3.11.

(I have a niggling worry about variance, but that shouldn't stop us from supporting this at runtime. Mypy doesn't support it yet anyway.)

File: Doc/library/typing.rst
Comment (JelleZijlstra) [inline Doc/library/typing.rst]:
```suggestion
      Added support for generic namedtuples.
```

**Review by JelleZijlstra — APPROVED** [review]
Looks good, just a wording suggestion.

I mentioned this at the typing summit today and people seemed positive. I also sent an email to typing-sig about it. Unless somebody brings up a compelling objection I'll merge this change before the feature freeze.

**Comment by kaylynn234** [comment]
> I probably agree, but I want to give type checker maintainers a chance to say "don't do this, it will introduce a giant soundness hole". I'll ask about it in the summit in a few hours.

I think asking for this kind of feedback is perfectly reasonable since the people maintaining type checkers tend to have a pretty good mental model of what is/isn't sound. That being said, I don't really like the idea of blocking a change like this - which would improve ergonomics a lot! - because a third-party tool **[1]** cannot/will not implement support for it. It really rubs me the wrong way.

**[1]**: I'm actually really confused about Mypy's status in this regard. It's under the Python org on GH but appears to still be considered third-party? I've seen a couple of proposals get shot down because of implementation concerns from Mypy now, and it's honestly started to bug me a lot. To my knowledge Python's static type system isn't controlled by Mypy (and exists more as a model for other type checking tools to pick up) so those decisions have always seemed really weird to me. If it's not too much to ask I would really appreciate some explanation here.

**Comment by JelleZijlstra** [comment]
@kaylynn234 it's not about mypy specifically; in fact, we know for a fact that mypy does *not* support this feature. There are other type checkers.

But the general idea is that `typing.py` exists to support the static type system, as specified in PEP-484 and subsequent PEPs. So we shouldn't add features to it that cannot be supported by static type checkers.

**Comment by gvanrossum** [comment]
> Before merging this PR I would like to review similar PR for TypedDict: #27663.

I like that we're doing this for TypedDict too, but why wait for that before merging this? I don't see how they're linked. @serhiy-storchaka @JelleZijlstra

**Review by JelleZijlstra — APPROVED** [review]

File: Doc/whatsnew/3.11.rst
Comment (JelleZijlstra) [inline Doc/whatsnew/3.11.rst]:
```suggestion
```

**Pull request merged**
